lib/typecheck.ts

```diff
@@ -392,1195 +421,1293 @@ async function typeCheckFilter(ast : Ast.BooleanExpression,
         }
-    }
 
-    for (const [, query] of Object.entries(klass.queries))
-        await typeCheckFunctionDef('query', query, schemas, useMeta);
-     for (const [, action] of Object.entries(klass.actions))
-        await typeCheckFunctionDef('action', action, schemas, useMeta);
-}
+        if (table instanceof Ast.VarRefTable)
+            this._addRequiredInputParamsInvocation(table, extrainparams);
+        else if (table instanceof Ast.InvocationTable)
+            this._addRequiredInputParamsInvocation(table.invocation, extrainparams);
+        else if (Utils.isUnaryTableToTableOp(table))
+            this._addRequiredInputParamsTable(table.table, extrainparams);
+        else
+            throw new TypeError();
+    }
 
-interface MixinDeclaration {
-    kind : string;
-    types : Type[];
-    args : string[];
-    required : boolean[];
-    is_input : boolean[];
-    facets : string[];
-}
+    private _addRequiredInputParamsAction(action : Ast.Action) {
+        if (action instanceof Ast.NotifyAction)
+            return;
 
-async function typeCheckMixin(import_stmt : Ast.MixinImportStmt,
-                              mixin : MixinDeclaration,
-                              schemas : SchemaRetriever,
-                              classes : ClassMap,
-                              useMeta : boolean) {
-    const presentParams = new Set<string>();
-    for (const in_param of import_stmt.in_params) {
-        const i = mixin.args.indexOf(in_param.name);
-        if (i === -1 || !mixin.is_input[i])
-            throw new TypeError(`Invalid parameter ${in_param.name} for mixin ${mixin.kind}`);
-        if (presentParams.has(in_param.name))
-            throw new TypeError(`Duplicate input parameter ${in_param.name}`);
-        presentParams.add(in_param.name);
-        if (in_param.value.isUndefined)
-            continue;
-        if (!in_param.value.isConstant())
-            throw new TypeError(`Mixin parameter ${in_param.name} must be a constant`);
-        const inParamType = mixin.types[i];
-        const valueType = await typeCheckValue(in_param.value, new Scope, schemas, classes, useMeta);
-        if (!Type.isAssignable(valueType, inParamType, {}, true))
-            throw new TypeError(`Invalid type for parameter ${in_param.name}, have ${valueType}, need ${inParamType}`);
-    }
-    for (let i = 0; i < mixin.args.length; i ++ ) {
-        if (mixin.required[i] && !presentParams.has(mixin.args[i]))
-            throw new TypeError(`Missing required parameter ${mixin.args[i]}`);
+        if (action instanceof Ast.VarRefAction)
+            this._addRequiredInputParamsInvocation(action, new Set<string>());
+        else if (action instanceof Ast.InvocationAction)
+            this._addRequiredInputParamsInvocation(action.invocation, new Set<string>());
+        else
+            throw new TypeError();
     }
-}
 
-function typeCheckMetadata(func : unknown) {
-    /*Object.entries(func.metadata).forEach(([name, value]) => {
```

Comment:
what happened to this?

Reply:
This code has been commented out fro a while, because it doesn't work with the latest format of canonical forms (with enums, argmin/max, etc.)